### PR TITLE
Fix Deployment with Recreate strategy not to wait on Pods in terminal phase

### LINKED ISSUE #60162: Deployment controller fails to create new RS when old RS has an evicted pod
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

The deployment controller for a `strategy: Recreate` deployment with a single replica failed to create a new ReplicaSet when the old ReplicaSet included an Evicted pod. Typically, the eviction was caused by "node low on resource: nodefs". When we manually deleted the evicted pod, the new RS and accompanying pod were created immediately. We have seen this happen at least a dozen times. Here are the audit logs for a recent case:

image

**What you expected to happen**:

The deployment controller ignores the evicted pod and creates the updated replicaset/pod.

**How to reproduce it (as minimally and precisely as possible)**:

To reproduce, you would need a way to trigger a kubelet-driven eviction on a node that is currently running a single-replica Deployment with  `strategy: Recreate`. Wait for the pod to be evicted and the new pod to be created, confirm that the evicted pod still exists, and then make a trivial change to the deployment spec. The running pod will be deleted, and no new replicaset/pod will be created. 

This issue does not seem to occur for manual evictions (via `POST /eviction`); pods I evict that way disappear from the API server immediately.

**Anything else we need to know?**:

Possibly related to https://github.com/kubernetes/kubernetes/issues/54525 and/or https://github.com/kubernetes/kubernetes/issues/55630

**Environment**:
- Kubernetes version (use `kubectl version`):  v1.8.5-gke.0
- Cloud provider or hardware configuration: GKE
- OS (e.g. from /etc/os-release): Container-Optimized OS

### PR DESCRIPTION
**Pull request by tnozicka**
**What this PR does / why we need it**:
Fixes Deployment with Recreate strategy not to wait on Pods in terminal phase. It can happen after eviction or failing to match selector and RS leaves such pod around right now. (Hopefully RC gets fixed separately.) 

**Which issue(s) this PR fixes** *:
Fixes https://github.com/kubernetes/kubernetes/issues/60162

**Special notes for your reviewer**:

**Release note**:

```release-note
Fixes a case when Deployment with recreate strategy could get stuck on old failed Pod.
```

/cc @janetkuo

### CODE REVIEW

**Comment by tnozicka** [comment]
Per @jsafrane: "volumes are unmounted for failed or succeeded pods" -  so we should be safe

**Comment by tnozicka** [comment]
tests are green :)

File: pkg/controller/deployment/recreate_test.go [hunk 1]
```diff
@@ -90,33 +90,165 @@ func TestOldPodsRunning(t *testing.T) {
...
-			name:     "old RSs without pods or non-zero status replicas",
-			oldRSs:   []*extensions.ReplicaSet{newRSWithStatus("rs-blabla", 0, 0, nil)},
-			expected: false,
+			name:              "old RSs without pods or non-zero status replicas",
+			oldRSs:            []*extensions.ReplicaSet{newRSWithStatus("rs-blabla", 0, 0, nil)},
+			hasOldPodsRunning: false,
+		},
+		{
+			name: "old RSs with zero status replicas but pods in terminal state are present",
+			oldRSs: []*extensions.ReplicaSet{
```
Comment (janetkuo) [inline pkg/controller/deployment/recreate_test.go]: Can use the same format as above test cases: `[]*extensions.ReplicaSet{newRSWithStatus("rs-blabla", 0, 0, nil)},`

File: pkg/controller/deployment/recreate.go [hunk 2]
```diff
@@ -104,8 +106,19 @@ func oldPodsRunning(newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSe
 		if newRS != nil && newRS.UID == rsUID {
 			continue
 		}
-		if len(podList.Items) > 0 {
-			return true
+		for _, pod := range podList.Items {
+			switch pod.Status.Phase {
+			case v1.PodFailed, v1.PodSucceeded:
```
Comment (janetkuo) [inline pkg/controller/deployment/recreate.go]: It should be impossible for RS to have succeeded pods, but there's no harm to add this check.

File: pkg/controller/deployment/recreate.go [hunk 3]
```diff
@@ -104,8 +106,19 @@ func oldPodsRunning(newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSe
...
+				continue
+			case v1.PodUnknown:
+				glog.Warning("Encountered Pod %s/%s in unknown state", pod.Namespace, pod.Name)
```
Comment (janetkuo) [inline pkg/controller/deployment/recreate.go]: IMO I don't think we need to check unknown pods and print a warning here. Just return `true` in default case should be enough.

**Comment by janetkuo** [comment]
Because ReplicaSets (and Deployments) only allow `restartPolicy:Always`, the only case their pods would end up in a terminated state is when 1) kubelet evicts a pod (causing the pod in a Failed state), or 2) kubelet rejects a pod when scheduling (also Failed pod). I confirmed with node team and it should be safe to assume that containers have terminated in both cases.

**Comment by janetkuo** [comment]
We should cherrypick this to 1.9.

File: pkg/controller/deployment/recreate.go (hunk 3 above)
Comment (kow3ns) [reply pkg/controller/deployment/recreate.go]: Agreed. That doesn't conform to the logging standards for controller_manager and could get spammy.

**Comment by kow3ns** [comment]
@tnozicka @janetkuo I'm assuming we want to target 1.10.0 with this and not cherry pick it back into 1.10.1 and 1.9.x ?

File: pkg/controller/deployment/recreate.go (hunk 2 above)
Comment (tnozicka) [reply pkg/controller/deployment/recreate.go]: It should be impossible, but a prefer to handle even such case explicitly.

File: pkg/controller/deployment/recreate_test.go (hunk 1 above)
Comment (tnozicka) [reply pkg/controller/deployment/recreate_test.go]: I prefer to state the tested value more verbosely in contrast to argument to a function that let's you look somewhere else for what the value means, but I have fixed it for consistency and your request.

File: pkg/controller/deployment/recreate.go (hunk 3 above)
Comment (tnozicka) [reply pkg/controller/deployment/recreate.go]: you are right - fixed, thx

**Comment by tnozicka** [comment]
Comments addressed.
I'll cherrypick it back to 1.9 once it merges.

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
> I'll cherrypick it back to 1.9 once it merges.

@tnozicka Thanks! Would you cherrypick it to 1.8 also?

**Pull request merged**